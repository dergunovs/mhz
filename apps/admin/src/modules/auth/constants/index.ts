export const API_POST_LOGIN = 'auth/login';
export const API_GET_CHECK_AUTH = 'auth/check';

export const URL_LOGIN = '/';
