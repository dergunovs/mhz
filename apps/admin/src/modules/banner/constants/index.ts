export const URL_BANNER = '/banners';
export const URL_BANNER_CREATE = `${URL_BANNER}/create`;
export const URL_BANNER_EDIT = `${URL_BANNER}/edit`;
