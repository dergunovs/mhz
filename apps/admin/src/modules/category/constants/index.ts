export const URL_CATEGORY = '/categories';
export const URL_CATEGORY_CREATE = `${URL_CATEGORY}/create`;
export const URL_CATEGORY_EDIT = `${URL_CATEGORY}/edit`;

export const CATEGORY_FIELD_TYPE_OPTIONS = ['string', 'number', 'boolean'];
