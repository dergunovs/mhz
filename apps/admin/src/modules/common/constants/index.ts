export const URL_MAIN = '/main';
export const URL_ERROR = '/404';
