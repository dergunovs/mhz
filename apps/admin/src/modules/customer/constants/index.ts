export const URL_CUSTOMER = '/customers';

export const API_CUSTOMER = 'customer';
