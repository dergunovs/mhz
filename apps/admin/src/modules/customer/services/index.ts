import { Ref, ComputedRef } from 'vue';

import { api, useQuery, IPageQuery, convertParams } from 'mhz-helpers';
import { API_CUSTOMER, ICustomer } from 'mhz-contracts';

export function getCustomers(query: Ref<IPageQuery | number>) {
  async function fn() {
    const params = convertParams(query);

    const { data } = await api.get<{ data: ICustomer[]; total: number }>(API_CUSTOMER, { params });

    return data;
  }

  return useQuery({ queryKey: [API_CUSTOMER, query], queryFn: fn });
}

export function getCustomer(id?: ComputedRef<string | string[]>) {
  async function fn() {
    if (!id?.value) return null;

    const { data } = await api.get<{ data: ICustomer }>(`${API_CUSTOMER}/${id.value}`);

    return data.data;
  }

  return useQuery({ queryKey: [API_CUSTOMER, id], queryFn: fn });
}
