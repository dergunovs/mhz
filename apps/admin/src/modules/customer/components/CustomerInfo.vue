<template>
  <div :class="$style.customer">
    <div>
      <b>Customer: </b>{{ props.customer.firstName }} {{ props.customer.lastName }}
      <a :href="`mailto:${props.customer.email}`">{{ props.customer.email }} </a>
    </div>

    <div><b>Created: </b>{{ formatDateTime(props.customer.dateCreated) }}</div>

    <div><b>Login date: </b>{{ formatDateTime(props.customer.dateLoggedIn) }}</div>

    <div v-if="props.customer.cart?.length">
      <div><b>Cart:</b></div>
      <div v-for="item in props.customer.cart" :key="item._id">
        {{ item.count }} x
        <RouterLink :to="`${URL_PRODUCT_EDIT}/${item.product._id}`">{{ item.product.title }}</RouterLink>
      </div>
    </div>

    <div v-if="props.customer.favouriteProducts?.length">
      <div><b>Favourites:</b></div>
      <div v-for="item in props.customer.favouriteProducts" :key="item._id">
        <RouterLink :to="`${URL_PRODUCT_EDIT}/${item._id}`">{{ item.title }}</RouterLink>
      </div>
    </div>

    <div v-if="props.customer.watchedProducts?.length">
      <div><b>Watched products:</b></div>
      <div v-for="item in props.customer.watchedProducts" :key="item._id">
        <RouterLink :to="`${URL_PRODUCT_EDIT}/${item.product._id}`">{{ item.product.title }}</RouterLink>
        {{ formatDateTime(item.dateCreated) }}
      </div>
    </div>
  </div>
</template>

<script setup lang="ts">
import { formatDateTime } from 'mhz-helpers';
import { ICustomer } from 'mhz-types';

import { URL_PRODUCT_EDIT } from '@/product/constants';

interface IProps {
  customer: ICustomer;
}

const props = defineProps<IProps>();
</script>

<style module lang="scss">
.customer {
  display: flex;
  flex-direction: column;
  gap: 8px;
}
</style>
