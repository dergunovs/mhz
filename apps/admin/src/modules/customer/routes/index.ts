import { URL_CUSTOMER } from '@/customer/constants';

export const customerRoutes = [
  {
    path: URL_CUSTOMER,
    name: 'CustomerList',
    component: () => import('@/customer/pages/CustomerListPage.vue'),
  },
  {
    path: `${URL_CUSTOMER}/:id`,
    name: 'Customer',
    component: () => import('@/customer/pages/CustomerPage.vue'),
  },
];
