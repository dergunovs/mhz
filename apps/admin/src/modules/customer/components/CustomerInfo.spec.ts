import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { formatDateTime } from 'mhz-helpers';

import CustomerInfo from './CustomerInfo.vue';

import { wrapperFactory } from '@/common/test';
import { CUSTOMERS } from '@/customer/fixtures';
import { URL_PRODUCT_EDIT } from '@/product/constants';

const CUSTOMER = CUSTOMERS.data[0];

const customerName = '[data-test="customer-info-name"]';
const customerEmail = '[data-test="customer-info-email"]';
const customerCreated = '[data-test="customer-info-created"]';
const customerLogin = '[data-test="customer-info-login"]';
const customerCart = '[data-test="customer-info-cart"]';
const customerCartItem = '[data-test="customer-info-cart-item"]';
const customerCartItemCount = '[data-test="customer-info-cart-item-count"]';
const customerCartItemTitle = '[data-test="customer-info-cart-item-title"]';
const customerFavourites = '[data-test="customer-info-favourites"]';
const customerFavouritesItem = '[data-test="customer-info-favourites-item"]';
const customerFavouritesTitle = '[data-test="customer-info-favourites-title"]';
const customerWatchedProducts = '[data-test="customer-info-watched-products"]';
const customerWatchedProductsItem = '[data-test="customer-info-watched-products-item"]';
const customerWatchedProductsDate = '[data-test="customer-info-watched-products-date"]';
const customerWatchedProductsTitle = '[data-test="customer-info-watched-products-title"]';

let wrapper: VueWrapper;

beforeEach(() => {
  wrapper = wrapperFactory(CustomerInfo, {
    props: {
      customer: CUSTOMER,
    },
  });
});

enableAutoUnmount(afterEach);

describe('CustomerInfo', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(CustomerInfo)).toBeTruthy();
  });

  it('shows customer name', async () => {
    expect(wrapper.find(customerName).text()).toEqual(`${CUSTOMER.firstName} ${CUSTOMER.lastName}`);
  });

  it('shows customer email', async () => {
    expect(wrapper.find(customerEmail).text()).toEqual(CUSTOMER.email);
    expect(wrapper.find(customerEmail).attributes('href')).toEqual(`mailto:${CUSTOMER.email}`);
  });

  it('shows date created', async () => {
    expect(wrapper.find(customerCreated).text()).toEqual(formatDateTime(CUSTOMER.dateCreated));
  });

  it('shows date logged in', async () => {
    expect(wrapper.find(customerLogin).text()).toEqual(formatDateTime(CUSTOMER.dateLoggedIn));
  });

  it('shows cart', async () => {
    expect(wrapper.find(customerCart).exists()).toEqual(true);

    await wrapper.setProps({ customer: CUSTOMERS.data[1] });

    expect(wrapper.find(customerCart).exists()).toEqual(false);
  });

  it('shows cart items', async () => {
    expect(wrapper.find(customerCart).exists()).toEqual(true);
    expect(wrapper.findAll(customerCartItem).length).toEqual(CUSTOMER.cart?.length);

    await wrapper.setProps({ customer: CUSTOMERS.data[1] });

    expect(wrapper.find(customerCart).exists()).toEqual(false);
  });

  it('shows cart item info', async () => {
    expect(wrapper.find(customerCartItemCount).text()).toEqual(CUSTOMER.cart?.[0].count.toString());
    expect(wrapper.find(customerCartItemTitle).text()).toEqual(CUSTOMER.cart?.[0].product.title);

    expect(wrapper.find(customerCartItemTitle).attributes('to')).toEqual(
      `${URL_PRODUCT_EDIT}/${CUSTOMER.cart?.[0].product._id}`
    );
  });

  it('shows favourites', async () => {
    expect(wrapper.find(customerFavourites).exists()).toEqual(true);
    expect(wrapper.findAll(customerFavouritesItem).length).toEqual(CUSTOMER.favouriteProducts?.length);

    await wrapper.setProps({ customer: CUSTOMERS.data[1] });

    expect(wrapper.find(customerFavourites).exists()).toEqual(false);
  });

  it('shows favourites info', async () => {
    expect(wrapper.find(customerFavouritesTitle).text()).toEqual(CUSTOMER.favouriteProducts?.[0].title);

    expect(wrapper.find(customerFavouritesTitle).attributes('to')).toEqual(
      `${URL_PRODUCT_EDIT}/${CUSTOMER.favouriteProducts?.[0]._id}`
    );
  });

  it('shows watched products', async () => {
    expect(wrapper.find(customerWatchedProducts).exists()).toEqual(true);
    expect(wrapper.findAll(customerWatchedProductsItem).length).toEqual(CUSTOMER.watchedProducts?.length);

    await wrapper.setProps({ customer: CUSTOMERS.data[1] });

    expect(wrapper.find(customerWatchedProducts).exists()).toEqual(false);
  });

  it('shows watched products info', async () => {
    expect(wrapper.find(customerWatchedProductsTitle).text()).toEqual(CUSTOMER.watchedProducts?.[0].product.title);
    expect(wrapper.find(customerWatchedProductsTitle).attributes('to')).toEqual(
      `${URL_PRODUCT_EDIT}/${CUSTOMER.watchedProducts?.[0].product._id}`
    );

    expect(wrapper.find(customerWatchedProductsDate).text()).toEqual(
      formatDateTime(CUSTOMER.watchedProducts?.[0].dateCreated)
    );
  });
});
