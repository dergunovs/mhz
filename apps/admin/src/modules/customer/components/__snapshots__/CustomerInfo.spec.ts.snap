// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomerInfo > matches snapshot 1`] = `
"<div class="_customer_2e0069">
  <div><b>Customer: </b><span data-test="customer-info-name">Alexandr Ivanot</span><a href="mailto:a@b.ru" data-test="customer-info-email">a@b.ru</a></div>
  <div><b>Created: </b><span data-test="customer-info-created">Sep 10, 2023, 2:24 PM</span></div>
  <div><b>Login date: </b><span data-test="customer-info-login">-</span></div>
  <div data-test="customer-info-cart">
    <div><b>Cart:</b></div>
    <div data-test="customer-info-cart-item"><span data-test="customer-info-cart-item-count">2</span> x <a to="/products/edit/64dd11c52f1f47a62b1cc0e2" data-test="customer-info-cart-item-title">ASUS ROG Strix Impact II Electro Punk</a></div>
    <div data-test="customer-info-cart-item"><span data-test="customer-info-cart-item-count">3</span> x <a to="/products/edit/64df860a7ac901ef3e27319c" data-test="customer-info-cart-item-title">Thermaltake TOUGHRAM 2x8 GB DDR4 4400</a></div>
  </div>
  <div data-test="customer-info-favourites">
    <div><b>Favourites:</b></div>
    <div data-test="customer-info-favourites-item"><a to="/products/edit/64dde049eb94d9f7131a41c9" data-test="customer-info-favourites-title">Razer BlackWidow V3 Mini</a></div>
    <div data-test="customer-info-favourites-item"><a to="/products/edit/64dd13d52f1f47a62b1cc18a" data-test="customer-info-favourites-title">ASUS ROG Gladius II Core</a></div>
    <div data-test="customer-info-favourites-item"><a to="/products/edit/64dd11c52f1f47a62b1cc0e2" data-test="customer-info-favourites-title">ASUS ROG Strix Impact II Electro Punk</a></div>
  </div>
  <div data-test="customer-info-watched-products">
    <div><b>Watched products:</b></div>
    <div data-test="customer-info-watched-products-item"><a to="/products/edit/64dddadbeb94d9f7131a40d0" data-test="customer-info-watched-products-title">ASUS ROG Azoth</a> - <span data-test="customer-info-watched-products-date">Sep 28, 2023, 3:38 PM</span></div>
    <div data-test="customer-info-watched-products-item"><a to="/products/edit/64dd11c52f1f47a62b1cc0e2" data-test="customer-info-watched-products-title">ASUS ROG Strix Impact II Electro Punk</a> - <span data-test="customer-info-watched-products-date">Oct 1, 2023, 8:15 AM</span></div>
  </div>
</div>"
`;
